pkg/finality-grandpa/lib.go

```diff
@@ -0,0 +1,398 @@
+func (cvr CommitValidationResult) NumEquiovcations() uint {
+	return cvr.numEquivocations
+}
+
+// NumInvalidVoters returns the number of invalid voters in the commit, i.e. votes from
+// identities that are not part of the voter set.
+func (cvr CommitValidationResult) NumInvalidVoters() uint {
+	return cvr.numInvalidVoters
+}
+
+// Validates a GRANDPA commit message.
+//
+// For a commit to be valid the round ghost is calculated using the precommits
+// in the commit message, making sure that it exists and that it is the same
+// as the commit target. The precommit with the lowest block number is used as
+// the round base.
+//
+// Signatures on precommits are assumed to have been checked.
+//
+// Duplicate votes or votes from voters not in the voter-set will be ignored,
+// but it is recommended for the caller of this function to remove those at
+// signature-verification time.
+func ValidateCommit[
+	Hash constraints.Ordered,
+	Number constraints.Unsigned,
+	Signature comparable,
+	ID constraints.Ordered,
+](
+	commit Commit[Hash, Number, Signature, ID],
+	voters VoterSet[ID],
+	chain Chain[Hash, Number],
+) (CommitValidationResult, error) {
+	validationResult := CommitValidationResult{
+		numPrecommits: uint(len(commit.Precommits)),
+	}
+
+	// filter any precommits by voters that are not part of the set
+	var validPrecommits []SignedPrecommit[Hash, Number, Signature, ID]
+	for _, signed := range commit.Precommits {
+		if !voters.Contains(signed.ID) {
+			validationResult.numInvalidVoters++
+			continue
+		}
+		validPrecommits = append(validPrecommits, signed)
+	}
+
+	// the base of the round should be the lowest block for which we can find a
+	// precommit (any vote would only have been accepted if it was targeting a
+	// block higher or equal to the round base)
+	var base HashNumber[Hash, Number]
+	var targets []HashNumber[Hash, Number]
+	for _, signed := range validPrecommits {
+		targets = append(targets, HashNumber[Hash, Number]{
+			Hash:   signed.Precommit.TargetHash,
+			Number: signed.Precommit.TargetNumber,
+		})
+	}
+	slices.SortFunc(targets, func(a HashNumber[Hash, Number], b HashNumber[Hash, Number]) bool {
+		return a.Number < b.Number
+	})
+	if len(targets) == 0 {
+		return validationResult, nil
+	}
+	base = targets[0]
+
+	// check that all precommits are for blocks that are equal to or descendants
+	// of the round base
+	var allPrecommitsHigherThanBase bool
+	for i, signed := range validPrecommits {
+		if chain.IsEqualOrDescendantOf(base.Hash, signed.Precommit.TargetHash) {
+			if i == len(validPrecommits)-1 {
+				allPrecommitsHigherThanBase = true
+			}
+			continue
+		}
+		break
+	}
+
+	if !allPrecommitsHigherThanBase {
+		return validationResult, nil
+	}
+
+	equivocated := &btree.Set[ID]{}
+
+	// add all precommits to the round with correct counting logic
+	round := NewRound[ID, Hash, Number, Signature](
+		RoundParams[ID, Hash, Number]{
+			RoundNumber: 0,
```

Comment:
new round number always is 0?

Reply:
This is for validating a commit.  It's just creating an arbitrary round to import the precommits into. I copied the `// doesn't matter here` comment over for this line.